numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -368,6 +368,400 @@ PyArray_GetSubType(int narrays, PyArrayObject **arrays) {
     return subtype;
 }
 
+/*
+ * Helper function for np.block to create error messages with depth info.
+ */
+void
+_concat_block_depth_str(PyObject **str, int depth, int *position){
+    int i;
+    PyUString_ConcatAndDel(str, PyUString_FromString("arrays"));
+    for (i = 0; i < depth; i++) {
+        PyObject *ind_str = PyUString_FromFormat("[%i]", position[i]);
+        PyUString_ConcatAndDel(str, ind_str);
+    }
+}
+
+/*
+ * Helper function for np.block which pads a shape with 1s.
+ */
+void
+_block_prepend_shape_ones(npy_intp *shape, int ndim, int n_ones){
+    int j;
+    for (j = ndim; j >= 0; j--) {
+        shape[j + n_ones] = shape[j];
+    }
+    for(j = 0; j < n_ones; j++) {
+        shape[j] = 1;
+    }
+}
+
+/*
+ * Helper function for np.block doing first pass to process input arrays. Given
+ * input of a nested list of ndarray/scalars, recursively compute the output
+ * ndim, dtype, shape, subtype, and also the max_depth of the input, and return
+ * a new nested list with all inner elements converted to ndarray.
+ *
+ * Returns a new reference to a nested list like the input nested list,
+ * converted to ndarrays. Sets dtype to a new reference, except on error in
+ * which case dtype will be null.
+ */
+PyObject *
+_get_block_info(PyObject *arrays, int *ndim, PyArray_Descr **dtype,
+                npy_intp *shape, double *sub_priority, PyTypeObject **subtype,
+                int *contigflag, int depth, int *concat_axis, int *position){
+
+    if (depth > NPY_MAXDIMS) {
+        PyErr_SetString(PyExc_ValueError, "arrays exceeded maximum dimensions");
+        return NULL;
+    }
+
+    *dtype = NULL;
+
+    if (PyTuple_Check(arrays)) {
```

Comment:
Is there a way to make this check a bit more robust, so that more is checked than just `tuple`?

Reply:
what kind of robustness are yuo thinking of adding? Can you point to a specific issue? This seems like the same logic as the all-python code.

Reply:
Currently, it only checks against `tuple`. Is it possible to check against other non-allowed structures in a more generic way? Something like `isinstance(x, list) and not isinstance(x, ndarray)`.

Reply:
I think it is because we also allow other array-like stuff. For convenience, we allow things like scalars

```
        # We've 'bottomed out' - arrays is either a scalar or an array
        size = _nx.size(arrays)
        return parent_index, _nx.ndim(arrays), size
```